### PR TITLE
Template args

### PR DESCRIPTION
**Pull request by Slava**
[Main Proposal](https://meteor.hackpad.com/Blaze-lexical-scope-and-template-arguments-fZP806qG6xQ)

A WIP PR and discussion for the template arguments. This branch is based on the `each-in` branch and `spacebars-template-exprs` branch.

```handlebars
<template name="myTemplate" args="fistName lastName">
  <p>{{firstName}} - {{lastName}}</p>
  <p>All arguments:</p>
  <ul>
    {{#each val in @args}}
        <li>{{@key}}: {{val}}</li>
    {{/each}}
  </ul>
</template>

<template name="main">
  {{@> myTemplate firstName="John" lastName="Doe" age=34 nationality="U.S."}}
</template>
```

Will output:

```html
<p>John - Doe</p>
<p>All arguments</p>
<ul>
  <li>firstName: John</li>
  <li>lastName: Doe</li>
  <li>age: 34</li>
  <li>nationality: U.S.</li>
</ul>
```

What's new:

1. The `@>` operator is the new inclusion with args (as opposed to `>` which is inclusion with the data context). Positional and keyword args are under consideration.
2. `@args` - special keyword to access a dictionary of passed arguments. Keyword args are mapped key-value, positional args are mapped as 0-arg0, 1-arg1. Iterable with `#each`.
3. Template's args signature: defined args will be dumped into scope for matching passed args. Other args, not present in the signature, will be available only in `@args`.
4. `@args` can be accessed in JS helpers as `Template.instance().args.get('x')` to get argument `x`. Similar to `Template.instance().state.get('x')/set('x', y)`.

Ported from https://github.com/meteor/meteor/pull/4118

### CODE REVIEW

File: packages/blaze-tools/tokens.js [hunk 1]
```diff
@@ -19,8 +19,14 @@ var unicodeClass = function (abbrev) {
 };
 
 // Special handlebars-reserved identifiers, usually starting with '@'
-// Other candidates to be in this list: '@last', '@first', '@key'
-var reservedIds = ['@index'];
+var reservedIds = [
+  // special #each-loop variables
+  '@index', '@index_1', '@first', '@last', '@odd', '@even', '@key',
```
Comment (RobertLowe) [inline packages/blaze-tools/tokens.js]: I guess `@index_1` and the others are just future proofing?

File: packages/blaze/template.js [hunk 2]
```diff
@@ -12,8 +12,14 @@
  * @locus Client
  * @param {String} [viewName] Optional.  A name for Views constructed by this Template.  See [`view.name`](#view_name).
  * @param {Function} renderFunction A function that returns [*renderable content*](#renderable_content).  This function is used as the `renderFunction` for Views constructed by this Template.
+ * @param {Object} [options]
+ * @param {String[]} [options.formalArgs] An array of names, the template
+ * accecpts as arguments. Passed arguments found in this list will have special
```
Comment (RobertLowe) [inline packages/blaze/template.js]: accecpts -> accepts

**Comment by RobertLowe** [comment]
A small note, but it seems a bit odd to me to have `args` and `state`. I mean, aren't `args` a form of state of a template. I guess for discussion and my general curiosity why not apply `args` into `state`? Is there some concern there?

**Comment by mitar** [comment]
+1 about why we have now already template data context, template instance, args, and state. Four places to put stuff.

**Comment by laosb** [comment]
It's much like a component instead of a template now.
Also much easier for package authors I think.
Data context for direct data, like `username`.
Args for sth need to query, like `userId`.

**Comment by RobertLowe** [comment]
A bit OT:

@laosb Then why not something like this:

`<component name='blogPostsForUser' userId='String' limit='Integer' randomness='Number' hideReadPosts='Boolean' checkByValidation='someMethod'>`

    Component.prototype.checkByValidation = function (checkByValidation) {
       check(checkByValidation, Match.Optional([String]))
     }

And enforce a protocol/interface... 

Just a thought...

**Comment by laosb** [comment]
@RobertLowe It's not clear enough for beginners.

**Comment by zimt28** [comment]
What's the status of this PR? It would be nice if it would be the other way around, defined args would only be accessible using `@args` and everything else would go into the scope, just like before.

**Comment by mitar** [comment]
Yea, I really do not like the direction this is going as well.

**Comment by tmeasday** [comment]
Thanks for this PR, sorry we never got around to merging it. We are currently cleaning up the repository and removing old PRs; if you think this PR still has value and are interested in resurrecting it, we encourage you to re-submit it against https://github.com/meteor/blaze

**Comment by Slava** [comment]
haha, it is ok! This PR was a bit too ambitious for its difficult time :-/

**Pull request closed without merging**